### ISSUE TITLE
Adjust field implementation

### ISSUE THREAD

**Issue by fermingut** [ArnauMiro/pyLowOrder#183]
As part of the CETACEO project, one of the objectives requested by Airbus is to interpolate the aerodynamic mesh Cp values onto a structural mesh while preserving one or more global quantities, such as CL or CD. Our proposed approach is to use one of the models available in `pyLOM.NN` (MLP, KANN, GNN) to predict Cp over each element of the structural mesh, and then apply an adjustment that minimally modifies the prediction so that the global CL and CD match those from the aerodynamic mesh.

This adjustment is formulated as a constrained optimization problem where the unknowns are the entire pressure field on the aerodynamic surface. I tested several methods to solve this problem. Due to the high number of variables (depending on the mesh density, currently ranging from 20,000 to 200,000 elements), classical optimization methods are too slow and memory-intensive, especially when building the Hessian matrix.
The best results were obtained using PyTorch’s Adam optimizer, which applies gradient descent to find the optimal solution.

Since this process needs to be part of pyLOM, I considered implementing a new method `adjust_field()` within the `pyLOM.Dataset` class. This method would allow adjusting a field according to a user-defined objective function. I’ve tried to design it in the most generic way possible so that it’s not limited to this specific use case, but can be applied to adjust any field based on any differentiable objective function.

Please let me know your thoughts. If you think this is too specific to include in `pyLOM.Dataset`, an alternative would be to implement a similar method in `pyLOM.NN.Dataset`. However, given how each dataset is structured, it seemed more natural and straightforward to implement it in `pyLOM.Dataset`.

And let me know if anything I’ve explained is unclear or needs further detail. 

```python 
def adjust_field(
        self,
        fieldname,
        obj_func,
        get_opt_param_func,
        constr_func=None,
        optimizer_class=torch.optim.Adam, 
        schduler_class=torch.optim.lr_scheduler.StepLR,
        opt_config=None,
        disp_progress=(False, 0),
        **kwargs
    ):
        default_config = {
            'niter': 1000,
            'lr': 1e-2,
            'lr_step_size': 1,
            'lr_gamma': 0.999,
            'penalty_factor': 1e5,
            'tolerance': 1e-9,
            'patience': 10,
        }

        if opt_config is None:
            opt_config = {}
        config = {**default_config, **opt_config}

        field = self[fieldname]
        field_mod = field.copy()
        field_losses = []

        def closure(colTensor, colTensor0, opt_vars):
            optimizer.zero_grad()
            obj_loss = obj_func(colTensor, colTensor0, **opt_vars)
            if constr_func is not None:
                penalty = constr_func(colTensor, **opt_vars)
                total_loss = obj_loss + config['penalty_factor'] * penalty
                total_loss.backward()
                return total_loss, obj_loss, penalty
            else:
                total_loss = obj_loss
                total_loss.backward()
                return total_loss, obj_loss, torch.tensor(0.0)

        for i, col in enumerate(field.T):
            colTensor = torch.tensor(col, requires_grad=True)
            colTensor0 = colTensor.clone().detach()
            opt_vars = get_opt_param_func(self, i, **kwargs)

            optimizer = optimizer_class([colTensor], lr=config['lr'])
            scheduler = schduler_class(optimizer, step_size=config['lr_step_size'], gamma=config['lr_gamma'])
            losses = []

            if disp_progress[0]:
                print(f"\nCase {i}:")

            prev_loss = float('inf')
            n_improvement = 0

            for epoch in range(config['niter']):
                total_loss, obj_loss, penalty = optimizer.step(lambda: closure(colTensor, colTensor0, opt_vars))
                scheduler.step()
                losses.append([obj_loss.item(), penalty.item(), total_loss.item()])

                if disp_progress[0] and (epoch % disp_progress[1] == 0):
                    print(f"Epoch {epoch:4}: Total Loss = {total_loss.item():.2e}, Objective = {obj_loss.item():.2e}, Penalty = {penalty.item():.2e}")

                loss_diff = abs(prev_loss - total_loss.item())
                prev_loss = total_loss.item()
                if loss_diff < config['tolerance']:
                    n_improvement += 1
                else:
                    n_improvement = 0
                if n_improvement >= config['patience']:
                    print(f"Early stopping at epoch {epoch}, no significant improvement.")
                    break
                if epoch >= config['niter'] - 1:
                    print(f"Reached maximum number of epochs ({config['niter']}). Stopping.")
                    break

            field_mod[:, i] = colTensor.detach().numpy()
            field_losses.append(losses)

        ndim = self.info(fieldname)['ndim']
        self.add_field(varname=fieldname + 'Adjusted', ndim=ndim, var=field_mod)
        return field_losses
```

**Comment by ArnauMiro** [comment]
Hi @fermingut,

Thanks for commenting this! I see a main problem here: pyLOM needs to be able to work without its ``NN`` module. What you are proposing is to implement features that use the ``N`` module outside of the ``NN`` module, thus will raise compatibility problems for sure.

Thus tool has to be inside the ``NN`` module, and I don't think it is appropriate to add it as part of the ``pyLOM.NN.Dataset`` class either. The way I suggest its implementation is the following:

1. We add a `interpolator.py` inside `pyLOM.NN` where you can implement your tool as a function.
2. Your function should take a `pyLOM.Dataset` as input (and possibly other necessary inputs from the source and target meshes?), and can/will return a new `pyLOM.Dataset` adapted to the new mesh.
3. Export the tool inside `pyLOM.NN.__init__.py` so that it is visible within the module.

Thus this will be called as `pyLOM.NN.adjust_field()` with the necessary inputs/outputs. I strongly suggest renaming this function to something that is clearer on what the tool does. If you are doing an interpolation between two meshes (now I'm thinking it might be useful for other situations than fluid to structural mesh), why not naming it `interpolation` or something of the sort?

Keep me posted on how the development goes!

**Comment by ArnauMiro** [comment]
@fermingut then it's fine, implement just the adjustment part of the process. It should be, nonetheless, within the `pyLOM.NN` module.

**Comment by fermingut** [comment]
I have added the changes to the branch 183-adjust_field.

I have created a class `Interpolator()` with the method `adjust_field()` that returns the `pyLOM.Dataset` with the new adjusted field. This way, an `interpolate()` method could also be added in the future if needed, to perform the first step of the interpolation.

Regarding the `adjust_field()` method, I have a question: Should the functions I am using for the adjustment also be included in pyLOM? For example, within the constraints function, I am using a function to calculate lift and drag. But that function is not currently implemented in pyLOM. What would be the most appropriate action? The functions are the following ones in my case:

```python
def objective(CoefPressureMod, CoefPressure, **kwargs):
    mse_loss = torch.nn.MSELoss(reduction='sum')
    return mse_loss(CoefPressureMod,CoefPressure)

def constraint_penalty(CoefPressureMod, alpha, normals, CL0, CD0, **kwargs):
    CL, CD = lift_drag_coeff_torch(alpha=alpha, normals=normals, CoefPressure=CoefPressureMod)
    penalty_CL = (CL - CL0)**2
    penalty_CD = (CD - CD0)**2
    factor = CL0/CD0
    return penalty_CL + factor * penalty_CD

def get_opt_param(dataset, i, **kwargs):
    alpha = torch.tensor(dataset.get_variable('aoa')[i] * torch.pi/180)
    normals = torch.tensor(dataset['Normals'].reshape(-1, 3))
    CL0 = torch.tensor(dataset.get_variable('CL')[i])
    CD0 = torch.tensor(dataset.get_variable('CD')[i])
    return {'alpha': alpha, 'normals': normals, 'CL0': CL0, 'CD0': CD0}

def lift_drag_coeff_torch(
    alpha,
    normals,
    Sref=80,
    CoefPressure=None,
    CoefSkinFriction=None,
):
    r"""
    Calculate lift and drag coefficients from pressure and skin friction coefficients.

    Args:
        alpha (torch.Tensor): Angle of attack in rad.
        normals (torch.Tensor): Tensor of shape (n, 3) with the normal vectors of the surface.
        Sref (float, optional): Reference area to calculate the lift and drag coefficients (default: ``80``).
        CoefPressure (torch.Tensor, optional): Tensor of shape (n,) or (n,1) with the pressure coefficients of the surface.
        CoefSkinFriction (torch.Tensor, optional): Tensor of shape (n,3) with the skin friction coefficients of the surface.

    Returns:
        Tuple[float, float]: Lift and drag coefficients.
    """
    CoefLiftPressure, CoefDragPressure, CoefLiftFriction, CoefDragFriction = 0, 0, 0, 0
    
    # Handle CoefPressure
    if CoefPressure is not None:
        if CoefPressure.ndimension() == 1:
            ForcePressure = - torch.matmul(CoefPressure, normals).flatten()
            CoefLiftPressure = ForcePressure[2]*torch.cos(alpha) - ForcePressure[0]*torch.sin(alpha)
            CoefDragPressure = ForcePressure[2]*torch.sin(alpha) + ForcePressure[0]*torch.cos(alpha)
        else:
            raise ValueError("CoefPressure must have shape (n,) or (n,1).")
    
    # Handle CoefSkinFriction
    if CoefSkinFriction is not None:
        if CoefSkinFriction.ndimension() == 1:
            CoefSkinFriction = CoefSkinFriction.view(-1, 3)
        elif CoefSkinFriction.shape[1] != 3:
            raise ValueError("CoefSkinFriction must have shape (n, 3).")
        
        Si = torch.sqrt(normals[:, 0]**2 + normals[:, 1]**2 + normals[:, 2]**2).view(-1, 1)
        ForceFriction = CoefSkinFriction * Si
        CoefLiftFriction = torch.sum(ForceFriction[:, 2] * torch.cos(alpha)) - torch.sum(ForceFriction[:, 0] * torch.sin(alpha))
        CoefDragFriction = torch.sum(ForceFriction[:, 2] * torch.sin(alpha)) + torch.sum(ForceFriction[:, 0] * torch.cos(alpha))
    
    if CoefPressure is None and CoefSkinFriction is None:
        raise ValueError("At least one of CoefPressure or CoefSkinFriction must be provided.")
    
    # Compute final lift and drag coefficients
    CL = (CoefLiftPressure + CoefLiftFriction) / Sref 
    CD = (CoefDragPressure + CoefDragFriction) / Sref 
    return CL, CD
```

**Comment by ArnauMiro** [comment]
Depends, we could add them too and try to give some flexibility to your tool so that a user could change them at a future stage. I am not sure if that makes sense for you.
